src/tests/ftest/util/dfuse_utils.py

```diff
@@ -25,6 +25,7 @@ def __init__(self, namespace, command):
         self.cuuid = FormattedParameter("--container {}")
         self.mount_dir = FormattedParameter("--mountpoint {}")
         self.sys_name = FormattedParameter("--sys-name {}")
+        self.thread_count = FormattedParameter("--thread-count {}")
```

Comment:
I've checked and this isn't in the doc so I'll add it in this PR as well.

Dfuse will by default run one thread per core it's allowed to run on, so if you don't do anything then that'll be one per core in the node.  You can use tasket or numactl (or anything that creates kernel process sets) to restrict dfuse to a subset of local cores, and if you do this then dfuse will still run one thread per core that it's allowed to run on.

As many dfuse operations block the calling thread whilst waiting for the server we may want to increase the thread count if the taskset is smaller in order to keep the number of operations-in-flight higher.  If you have 128 cores then that limits you to 128 concurrent operations (read and write are asynchronous so this is metadata operations only), but once you start getting onto smaller core counts then you want to increase the ratio to avoid limiting the throughput.  For simple benchmarks like fio then you want at least one thread per fio process for example.

In this test it'll run with a 2:1 ratio as you say which is probably a sensible number to start with.

In practice getting the taskset correct is going to be tricky, any MPI job will likely already be bound to cores for example, and binding dfuse to the same cores would have a negative effect, so you may want to bind dfuse to cores 40-43 and tell mpirun to not schedule anything to these.  I'm not sure what you mean by "job -ppn", is the the mpirun option?

Reply:
yes it is a mpiexec.hydra option(mpich);  -ppn is the number of mpi processes to assign per node (-np => ppn*num of nodes).     It is similar to orterun slots used in the hostfile

Reply:
That's what I thought.  You need to read https://wiki.mpich.org/mpich/index.php/Using_the_Hydra_Process_Manager#Process-core_Binding

Back to your example, if the client has 44 cores then you might want to dedicate 4 cores to dfuse and 40 to the app, in which case you'd want to overcommit threads to cores in dfuse so you might choose to use cores 40-43 for dfuse with --thread-count 8, then -ppn 40 for the app, but you'll probably need hydra options, "-bind-to core" might be enough, and that may be the default but you'll need to check as per the link above.

Have a look at the taskset options as well, as you may find that 36-43:2 or similar might be better.

Reply:
right, I understand how the binding works for mpi; and -bind-to core is probably all I will need to add and based on what I see when the jobs run it does appear to be the default but I will still add it.    I was concerned with the relationship between the number of process per node and the number of cores passed to taskset.   I will take this offline; I have questions about the implementation; not about the PR